core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp

```diff
@@ -0,0 +1,356 @@
+#include "common.h"
+#include <string>
+#include <string.h>
+#include <stdlib.h>
+
+// string functions
+
+extern "C" DLL_EXPORT int STDMETHODCALLTYPE CountUtf8Characters(char* value)
+{
+    // value is ANSI (on Windows)/ UTF8 (on Linux/Mac) null-terminated string
+    // and can be null.
+    if (value == nullptr)
+    {
+        return -1;
+    }
+    else
+    {
+        return strlen(value);
+    }
+}
+
+extern "C" DLL_EXPORT int STDMETHODCALLTYPE CountUtf16Characters(WCHAR* value)
+{
+    // value is UTF16 null-terminated string
+    // and can be null.
+    if (value == nullptr)
+    {
+        return -1;
+    }
+    else
+    {
+        return std::char_traits<WCHAR>::length(value);
+    }
+}
+
+extern "C" DLL_EXPORT int STDMETHODCALLTYPE CountPlatformSpecificCharacters(char_t* value)
+{
+    // char_t is defined as WCHAR on Windows and as char (UTF8) on Linux/Mac
+    if (value == nullptr)
+    {
+        return -1;
+    }
+    else
+    {
+#ifdef WINDOWS
+        return wcslen(value);
+#else
+        return strlen(value);
+#endif
+    }
+}
+
+char* CreateCopyOfString(const char* stringValue, int& stringValueSize)
+{
+    stringValueSize = strlen(stringValue) + 1; // + 1 for null terminator
+
+#ifdef WINDOWS
+    // On Windows the returned buffer is going to be freed with CoTaskMemFree, so it needs to be allocated
+    // with CoTaskMemAlloc or CoTaskMemRealloc.
+    char* buffer = (char*)CoTaskMemAlloc(stringValueSize);
+#else
+    // On Linux and Mac the returned buffer is going to be freed with free(), so it needs to be allocated
+    // with malloc/callos/realloc.
+    char* buffer = (char*)malloc(stringValueSize);
+#endif
+
+    if (buffer == nullptr)
+    {
+        return nullptr;
+    }
+
+    memcpy(buffer, stringValue, stringValueSize);
+    return buffer;
+}
+
+extern "C" DLL_EXPORT int STDMETHODCALLTYPE GetStringIntoCalleeAllocatedBuffer(char ** value)
+{
+    const char* stringValue = "Native string value - callee allocated";
```

Comment:
`callee` maybe is `caller`

Reply:
`callee` is correct - the string buffer is allocated by the native function.